SQL DDL prototype

gierlachg opened the pull request.

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/571/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/571/Hazelcast-pr-builder-571.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 8 errors reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java:151: Don't use trailing comments. [TrailingComment]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/sql/impl/connector/SqlConnectorFactory.java:23:1: Utility classes should not have a public or default constructor. [HideUtilityClassConstructor]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/sql/impl/parser/NoOpSqlParser.java:1: Missing a header - not enough lines in file. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/sql/impl/parser/SqlParseTask.java:1: Line does not match expected header line of '/*'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/sql/impl/parser/Statement.java:1: Missing a header - not enough lines in file. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/sql/impl/schema/Catalog.java:35: Don't use trailing comments. [TrailingComment]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/sql/impl/schema/TableSchema.java:1: Line does not match expected header line of '/*'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/ReplicatedMapTableResolver.java:69: Don't use trailing comments. [TrailingComment]
--------------------------
</details>
</pre>

viliam-durina commented on hazelcast-sql/src/main/codegen/includes/parserImpls.ftl:
```diff
@@ -0,0 +1,312 @@
+                    <LPAREN> variant = SimpleIdentifier() <RPAREN>
+                ]
+                {
+                    if (variant == null) {
```
```suggestion
                    if (variant == null || "OFFSET_DATE_TIME".equals(variant.getSimple()) {
```

viliam-durina commented on hazelcast-sql/src/main/codegen/includes/parserImpls.ftl:
```diff
@@ -0,0 +1,312 @@
+                    <LPAREN> variant = SimpleIdentifier() <RPAREN>
+                ]
+                {
+                    if (variant == null) {
```
```suggestion
                    if (variant == null || "INSTANT".equals(variant.getSimple()) {
```

viliam-durina commented on hazelcast-sql/src/main/codegen/includes/parserImpls.ftl:
```diff
@@ -0,0 +1,312 @@
+                }
+            )
+        |
+            <WITHOUT> <TIME> <ZONE> { type = QueryDataType.TIMESTAMP; }
```
This should be also the default when just `TIMESTAMP` is used without any TIME ZONE clause

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/575/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/575/Hazelcast-pr-builder-575.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 2 errors reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/sql/impl/connector/SqlConnectorFactory.java:23: Class SqlConnectorFactory should be declared as final. [FinalClass]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast/src/main/java/com/hazelcast/sql/impl/connector/SqlConnectorFactory.java:36:5: Constructor definition in wrong order. [DeclarationOrder]
--------------------------
</details>
</pre>

devozerov commented on hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/optimizer/SqlOptimizer.java:
```diff
@@ -14,77 +14,50 @@
     private final NodeEngine nodeEngine;
 
-    public CalciteSqlOptimizer(NodeEngine nodeEngine) {
+    public SqlOptimizer(NodeEngine nodeEngine) {
```
I propose to avoid renames of the central classes unless there is a strong reason to. In this specific case, parsing is a part of optimization process, not vice versa.

gierlachg replied:
I can avoid renames but it is going to be awkward.
Regarding what is part of what. I have extracted parsing part from `CalciteSqlOptimizer`, there is nothing to optimize in DDL statements, thus created `SqlParser` that is able to convert a sql string into an executable statement, part of that conversion for DQLs is optimization. Not sure how else it could be done.

devozerov commented on hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/schema/HazelcastSchemaUtils.java:
```diff
@@ -65,30 +68,33 @@ public static HazelcastSchema createRootSchema(List<TableResolver> tableResolver
 
-                Map<String , HazelcastTable> schemaTableMap =
-                    tableMap.computeIfAbsent(table.getSchemaName(), (k) -> new HashMap<>());
+                if (table.getSchemaName() == null || table.getSchemaName().isEmpty()) {
```
We cannot use empty schema names. A classical object hierarchy is `catalog -> schema -> object`. Every such an object should have a definitive name for several reasons:
1. They need to be displayed somehow in tools that will talk to us through JDBC
1. It should be possible to reach any object from any naming context. With empty schema name, some objects may become unreachable

Other vendors typically have a default schema name `PUBLIC` or alike.

devozerov commented on hazelcast/src/main/java/com/hazelcast/sql/SqlService.java:
```diff
@@ -49,4 +50,22 @@ default SqlCursor query(String sql, Object... params) {
+     *
+     * @param sql SQL.
+     */
+    default void update(String sql) {
```
I am not very keen on having such methods on public API because it increases complexity for users and we need a common method that handles all query types anyway, because otherwise, we will not be able to integrate JDBC. 
Is there a reason why the existing method or at least re-use of `SqlQuery` is not applicable for an update operation?

gierlachg replied:
If we reused existing method `SqlCursor query(SqlQuery query)` we would need to return `SqlCursor` from an execution of DDL statement, of course I could return a cursor containing single value of `0` but it just felt cumbersome - that's the thing we were discussing yesterday - I can squeeze DDL into existing method for now if we don't want to make any decision now and postpone the API decision till later.

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/576/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/576/Hazelcast-pr-builder-576.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast-sql: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 66 errors reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/parser/ParserResource.java:1: File does not end with a newline. [NewlineAtEndOfFile]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/parser/SqlDataType.java:37:5: Cyclomatic Complexity is 20 (max allowed is 12). [CyclomaticComplexity]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:1: Line does not match expected header line of '/*'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:96:67: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:97:50: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:98:43: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:99:45: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:100:46: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:101:43: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:102:45: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:103:38: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:104:42: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:105:41: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:106:44: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:107:51: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:108:54: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:109:48: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:110:44: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:111:53: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:112:80: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:113:82: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:114:72: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:115:90: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/SchemaTest.java:116:68: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:61:12: More than 10 parameters (found 19). [ParameterNumber]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:183:33: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:184:44: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:185:46: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:186:42: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:187:36: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:188:38: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:189:34: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:190:36: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:191:38: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:192:40: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:193:46: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:194:46: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:195:44: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:196:44: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:197:52: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:198:34: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:199:42: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:200:40: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:201:52: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:202:54: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:217:46: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:218:43: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:219:37: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:220:39: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:221:35: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:222:37: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:223:57: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:224:60: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:225:53: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:226:61: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:227:61: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:228:59: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:229:59: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:230:67: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:231:49: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:232:57: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:233:55: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:234:67: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/Person.java:1: File does not end with a newline. [NewlineAtEndOfFile]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/Person.java:1: Line does not match expected header line of '/*'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_2/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/Person.java:46:50: '&&' should be on a new line. [OperatorWrap]
--------------------------
</details>
</pre>

devozerov commented on hazelcast/src/main/java/com/hazelcast/sql/impl/SqlServiceImpl.java:
```diff
@@ -156,84 +164,106 @@ private SqlCursor query0(String sql, List<Object> params, long timeout, int page
+    }
 
-            state = internalService.executeExplain(plan);
+    private DqlStatement prepareQuery(String sql) {
```
It is not clear why SELECT is wrapped into a DDL class

gierlachg replied:
I'm not sure I follow.

devozerov commented on hazelcast/src/main/java/com/hazelcast/sql/impl/parser/SqlParser.java:
```diff
@@ -14,19 +14,18 @@
+ * Parser responsible for conversion of SQL string to executable statement.
  */
-public interface SqlOptimizer {
+public interface SqlParser {
```
As mentioned before, the motivation of the rename is not obvious. This underlying implementation is optimizer, not parser.

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/579/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/579/Hazelcast-pr-builder-579.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast-sql: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-sql/src/test/java/com/hazelcast/sql/schema/model/AllTypesValue.java:61:12: More than 10 parameters (found 19). [ParameterNumber]
--------------------------
</details>
</pre>

gierlachg commented:
So I have refactored everything to reuse `query`, reverted renames, squeezed DDL into existing optimizer... in general as few changes as possible -> I can't say I like how `SqlServiceImpl` & `CalciteSqlOptimizer` look like but it works.

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/580/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/580/Hazelcast-pr-builder-580.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal com.github.spotbugs:spotbugs-maven-plugin:3.1.12:spotbugs (spotbugs) on project hazelcast-sql: Execution spotbugs of goal com.github.spotbugs:spotbugs-maven-plugin:3.1.12:spotbugs failed: Java returned: 1 -> [Help 1]
--------------------------
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/583/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/583/Hazelcast-pr-builder-583.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientDynamicConfigurationAwareConfigTest>DynamicConfigurationAwareConfigTest.testDecorateAllPublicMethodsFromTest:61 Method public com.hazelcast.config.SqlConfig com.hazelcast.config.Config.getSqlConfig() is declared by class com.hazelcast.config.Config whilst it should be declared by class com.hazelcast.client.impl.clientside.ClientDynamicClusterConfig
[ERROR]   DynamicConfigurationAwareConfigTest.testDecorateAllPublicMethodsFromTest:61 Method public com.hazelcast.config.SqlConfig com.hazelcast.config.Config.getSqlConfig() is declared by class com.hazelcast.config.Config whilst it should be declared by class com.hazelcast.internal.dynamicconfig.DynamicConfigurationAwareConfig
[ERROR] Errors: 
[ERROR]   DataSerializableImplementsVersionedTest.testDataSerializableForVersionedInterface ? StackOverflow
[INFO] 
[ERROR] Tests run: 27742, Failures: 2, Errors: 1, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

gierlachg commented:
verify

gierlachg commented:
run-lab-run

devozerov commented on hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/parse/SqlDataType.java:
```diff
@@ -0,0 +1,80 @@
+    @Override
+    @SuppressWarnings("checkstyle:CyclomaticComplexity")
+    public void unparse(SqlWriter writer, int leftPrec, int rightPrec) {
+        if (type == QueryDataType.BOOLEAN) {
```
Minor thought for future: maybe it makes sense to add this "unparsing" to QueryDataType itself so that we do not miss something when a new type is added. 
Not relevant for prototyping.

devozerov commented on hazelcast/src/main/java/com/hazelcast/sql/impl/SqlServiceImpl.java:
```diff
@@ -174,10 +174,16 @@ private SqlCursor query0(String sql, List<Object> params, long timeout, int page
             state = internalService.executeExplain(plan);
         } else {
             Plan plan = prepare(sql);
+            if (plan == null) {
```
Here we make the purpose of optimization stage fuzzy, because sometimes optimizer does optimization, sometimes it does parsing and execution. I would prefer to retain clear boundaries for component responsibilities so that the optimizer prepares the plan, and then in the `SqlServiceImpl` we decide how to execute the returned plan. This is needed to ensure that we can evolve the engine easily. For example, consider that in future we will need to implement SQL query statistics or asynchronous query cancellation for JDBC. With a centralized approach, it will be much easier than trying to hook into disparate execution points. 

I do not want to slow down the prototyping due to this comment, so I propose to do the following: 
1) Go on with the current approach for now
2) Soon I will switch to API implementation, where I can try to decouple execution from planning in the following way:
- Planner return a plan that could be SELECT plan, DDL plan, etc
- Then depending on the plan type, we feed it to a designated handler in the "core" module

gierlachg commented on hazelcast/src/main/java/com/hazelcast/sql/impl/connector/SqlConnectorFactory.java:
```diff
@@ -0,0 +1,42 @@
+    private SqlConnectorFactory() {
+    }
+
+    private static Map<String, SqlConnector> prepareConnectors() {
```
That is yet to be decided, we might pass the class name as `type` while creating the table so we could even accept user defined connectors, however that would require publishing certain classes that are private now - `Table`, `QueryDataType` etc.

devozerov replied:
I see. Given that we do not expect more serious code conflicts around the optimizer code base for now, it may worth trying to add some real connectors (say, Kafka) to see how it to wire it up properly. Having such connectors not only will help us get a better understanding on how to wire them up but also will help us get rid of connectors for cluster-local maps. 
But maybe it is better to do this separately, to keep PRs more manageable.

gierlachg commented on hazelcast/src/main/java/com/hazelcast/sql/impl/connector/SqlConnector.java:
```diff
@@ -0,0 +1,33 @@
+    Table createTable(
+            String schemaName,
+            String name,
+            List<Field> fields,
```
You mean something else than `Map<String, String> options` ?

devozerov replied:
Sorry, please disregard. I meant field options, not table options. And field options are already present.

gierlachg commented on hazelcast/src/main/java/com/hazelcast/sql/impl/connector/PartitionedMapConnector.java:
```diff
@@ -0,0 +1,54 @@
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+
+// TODO: do we want to keep it? maps are auto discovered...
```
What about leaving it till we we have some tests on Jet side using IMDG `SqlService`?

devozerov replied:
Could you please clarify what kind of tests you mean? I am not against keeping it for a while, but please note that this connector creates a table that does not expose useful map information, such as indexes and distribution field. Therefore its usage even in tests is questionable.

gierlachg replied:
Something like `SchemaTest` so we make sure DDL/creating tables works and it is not broken during development of other things.

devozerov replied:
Got it, thanks.

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/588/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/588/Hazelcast-pr-builder-588.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientDynamicConfigurationAwareConfigTest>DynamicConfigurationAwareConfigTest.testDecorateAllPublicMethodsFromTest:61 Method public com.hazelcast.config.SqlConfig com.hazelcast.config.Config.getSqlConfig() is declared by class com.hazelcast.config.Config whilst it should be declared by class com.hazelcast.client.impl.clientside.ClientDynamicClusterConfig
[ERROR]   DynamicConfigurationAwareConfigTest.testDecorateAllPublicMethodsFromTest:61 Method public com.hazelcast.config.SqlConfig com.hazelcast.config.Config.getSqlConfig() is declared by class com.hazelcast.config.Config whilst it should be declared by class com.hazelcast.internal.dynamicconfig.DynamicConfigurationAwareConfig
[ERROR] Errors: 
[ERROR]   DataSerializableImplementsVersionedTest.testDataSerializableForVersionedInterface ? StackOverflow
[INFO] 
[ERROR] Tests run: 27742, Failures: 2, Errors: 1, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/589/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/589/Hazelcast-pr-builder-589.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientDynamicConfigurationAwareConfigTest>DynamicConfigurationAwareConfigTest.testDecorateAllPublicMethodsFromTest:61 Method public com.hazelcast.config.SqlConfig com.hazelcast.config.Config.getSqlConfig() is declared by class com.hazelcast.config.Config whilst it should be declared by class com.hazelcast.client.impl.clientside.ClientDynamicClusterConfig
[ERROR]   DynamicConfigurationAwareConfigTest.testDecorateAllPublicMethodsFromTest:61 Method public com.hazelcast.config.SqlConfig com.hazelcast.config.Config.getSqlConfig() is declared by class com.hazelcast.config.Config whilst it should be declared by class com.hazelcast.internal.dynamicconfig.DynamicConfigurationAwareConfig
[ERROR] Errors: 
[ERROR]   DataSerializableImplementsVersionedTest.testDataSerializableForVersionedInterface ? StackOverflow
[INFO] 
[ERROR] Tests run: 27742, Failures: 2, Errors: 1, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>
